[#8] feat: 좌석 구역(SeatBlock) 도메인 개발

jn4624 opened the pull request:
## 작업 내용
1. 도메인 개발
    - SeatBlock과 Stadium의 연관관계 설정.
        - 경기장마다 구역 구성이 다르므로 별도 관리.
        - 구역에 할당할 좌석 수가 경기장의 총 수용 인원을 초과하지 않도록 제한.
## 기술적 고려 사항
1. 경기장별 구역 레벨과 가격 관리 방식 검토.
    - 경기장마다 구역 레벨과 가격도 다르므로 단순한 필드로 관리 불가능.
2. 시즌별 가격 변동 로직 적용 검토.
    - 변동 가격을 관리하기보다는 가격의 상승률을 관리하는 것이 효율적인지 검토.
    - 서비스 레이어에서 동적으로 처리하는 방안 검토.
## 향후 개선 방향
1. 명확한 테스트 네이밍 컨벤션 필요.
2. 예외 글로벌 처리 필요.

jn4624 committed:
[#8] feat: 좌석 구역(SeatBlock) 도메인 개발

jn4624 committed:
[#8] feat: 응답 객체 분리

roam-f-lab commented on src/main/java/com/kbo/seatblock/controller/request/SeatBlockRequest.java:
```diff
@@ -0,0 +1,19 @@
+package com.kbo.seatblock.controller.request;
+
+public class SeatBlockRequest {
```
record를 사용해 볼 수도 있을 것 같습니다.

jn4624 replied:
Record 클래스를 사용하는 방안도 고려했지만, 자바 16부터 정식 기능으로 채택되었기 때문에 의존성을 고려하여 사용하지 않았습니다.
하지만 Record 클래스가 제공하는 자동 생성 기능과 불변성을 활용하는 것이 더 효율적이라 생각되어 Record 클래스를 적용하는 방향으로 수정 작업을 진행하겠습니다.

roam-f-lab commented on src/main/java/com/kbo/stadium/entity/Stadium.java:
```diff
@@ -41,4 +43,15 @@ public String getName() {
+	}
+
+	public void allocateSeatCount(int seatCount) {
+		if ((allocatedSeatCount + seatCount) > this.capacity) {
```
수용 인원이 21억이 되지는 않겠지만, 코드 구성이 오버플로우가 발생하면 오동작할 수 있습니다.

jn4624 replied:
수용 인원의 범위가 크지 않을 것으로 예상하여 int 자료형을 사용했으나, 오버플로우 가능성을 고려하여 long 자료형으로 수정하겠습니다.

roam-f-lab replied:
수용인원 이라는 정상 데이터 범위에서 int의 범위를 초과하는 케이스가 없겠지만, 제가 언급한 오버플로우는 allocatedSeatCount + seatCount 이 연산에서 음수가 나오는 케이스로 정상 데이터가 아닐 때 발생할 수 있는 예외 케이스를 언급했습니다.
long이어도 같은 로직을 유지한다면 정수 오버플로우는 발생할 수 있습니다.

jn4624 replied:
allocatedSeatCount + seatCount 연산이 long의 최대값을 초과하는지 미리 확인한 후 연산을 수행하는 방향으로 가야 할까요?

roam-f-lab replied:
일단, long으로 확장한 타입을 다시 int로 내리면 아래와 같이 처리가 가능합니다.
```
public void allocateSeatCount(int seatCount) {
  if (seatCount < 0) {
    에러 처리
  }
  if (((long) allocatedSeatCount + seatCount) > this.capacity) {
    throw new IllegalStateException("Stadium total capacity exceeded.");
  }
  allocatedSeatCount += seatCount;
}
```

jn4624 replied:
덧셈 연산을 long 범위에서 수행하게 만들어서 int의 오버플로우를 방지하는 방법일까요?
전혀 생각해보지 못했습니다 😯
처음부터 문제에 대한 접근이 잘못되었네요 🥲 수정하도록 하겠습니다.

jn4624 committed:
[#8] feat: PR 리뷰 반영

jn4624 committed:
[#8] feat: PR 리뷰 반영

- int 오버플로우 방지

jn4624 committed:
[#8] feat: 연관관계 매핑 수정 및 좌석 구역 리스트 기능 제거

- 좌석 구역 리스트는 Stadium을 통해 제공하는 것으로 설계를 변경하여 SeatBlock이 제공하던 기능 제거
- Stadium과 SeatBlock의 연관관계를 단방향에서 양방향으로 수정
